### PR TITLE
Improve terminology validation error messages

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java [hunk 1]
```diff
@@ -197,9 +199,11 @@ public CodeValidationResult validateCode(
 				}
 			}
 
-			return new CodeValidationResult()
-					.setSeverity(IssueSeverity.ERROR)
-					.setMessage("Code \"" + theCode + "\" is not in system: " + USPS_CODESYSTEM_URL);
+			String actualSystem = defaultIfBlank(theCodeSystem, expectSystem);
```
Comment (jmarchionatto) [inline hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:202]: thought: I know is not part of your changes but this code become pretty "complex" (in the sense of "A Philosophy of Software Design" book)

**Comment by jmarchionatto** [comment]
Looks good! Approved with a few comments

**Review by jmarchionatto — APPROVED** [review]
Looks Good! Approved with a few comments

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/util/OperationOutcomeUtil.java [hunk 2]
```diff
@@ -250,34 +251,77 @@ public static void addLocationToIssue(FhirContext theContext, IBase theIssue, St
...
+			ExtensionUtil.setExtension(
+					theCtx,
+					theIssue,
+					"http://hl7.org/fhir/StructureDefinition/operationoutcome-issue-line",
+					"integer",
+					theLine);
+		}
+	}
+
+	public static void addIssueColExtensionToIssue(FhirContext theCtx, IBase theIssue, String theLine) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/util/OperationOutcomeUtil.java:306]: oops fixed!

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/ValidationResult.java [hunk 3]
```diff
@@ -171,6 +167,44 @@ public void populateOperationOutcome(IBaseOperationOutcome theOperationOutcome)
...
+			String messageId) {
+		if (isBlank(location) && locationLine != null && locationCol != null) {
+			location = "Line[" + locationLine + "] Col[" + locationCol + "]";
+		}
+		String severity = issueSeverity != null ? issueSeverity.getCode() : null;
+		IBase issue = OperationOutcomeUtil.addIssueWithMessageId(
+				myCtx, theOperationOutcome, severity, message, messageId, location, Constants.OO_INFOSTATUS_PROCESSING);
+
+		if (locationLine != null || locationCol != null) {
+			String unknown = "(unknown)";
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/ValidationResult.java:186]: done

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/ValidationResult.java [hunk 4]
```diff
@@ -131,37 +131,33 @@ public IBaseOperationOutcome toOperationOutcome() {
...
+							locationLine,
+							locationCol,
+							issueSeverity,
+							combinedMessage,
+							messageId);
 				}
+			} else {
+				addIssueToOperationOutcome(
+						theOperationOutcome, location, locationLine, locationCol, issueSeverity, message, messageId);
 			}
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/ValidationResult.java:161]: done

File: hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/support/CommonCodeSystemsTerminologyService.java:202]: agree..... although i actually think this refactor simplified it a tiny bit

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/SingleValidationMessage.java [hunk 5]
```diff
@@ -32,6 +34,7 @@ public class SingleValidationMessage {
 	private String myMessage;
 	private String myMessageId;
 	private ResultSeverityEnum mySeverity;
+	private List<String> mySliceMessages;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/validation/SingleValidationMessage.java:37]: oops good call, fixed

**Pull request merged**